### ISSUE TITLE
autocomplete on table fields stopped working

### ISSUE THREAD

**Issue by anselal** [HeidiSQL/HeidiSQL#917]
## Steps to reproduce this issue

1. Start typing a query
2. assign an alias to a table eg. d
3. type d. or press control+space
4. Then I get nothing

## Current behavior

autocomplete on table fields does not work

## Expected behavior

It should display a popup with the names of field of the specific table

## Possible solution

revert a commit ???

## Environment

* **HeidiSQL version:** 10.3.0.5901
* Win 10 Pro 64bit
* MySQL 5.5.62

**Comment by ansgarbecker** [comment]
This is due to pull request #914 for issue for issue #895 , so let us fix it there. Thanks however for the quick feedback... just a few hours after committing the related change :)

**Comment by anselal** [comment]
No problem. I am using HeidiSQL everyday so issues like this are important to be fixed.
thanks for the great work

**Comment by zilzila** [comment]
@anselal, could you please provide a test query with the name of the table and database. Is the database in the tree on the left selected?

**Comment by zilzila** [comment]
I had in mind a test query, not a real one, to understand what type of identifiers are used (e.g. CamelCased or \`quoted\`).
For example, a query like `select * from table_name d where d.` works fine.
I suppose the query was like `select * from TableName d where d.` or `select d. from TableName d`, where were the CamelCase identifiers, right?

**Comment by anselal** [comment]
Oh, sorry I didn't get that.
the queries I write are always lowercase like `select * from table_name d where d.`

**Comment by zilzila** [comment]
In this test, I have autocomletion popping up:
```sql
-- drop table my_test_table;
create table my_test_table (id int);
```
Refresh current database in DBTree (on the left).
```sql
select * from my_test_table d where d. 
```
> Any query fails to autocomplete the table fields except in the `incormation_schema` database.

Maybe the database `information_schema` in DBTree was selected?

**Comment by anselal** [comment]
no, i have selected the correct database, else the query `select * from _test_table` would not run.

**Comment by anselal** [comment]
I tried your example and it works. I think that the problem is that my table names start with an Underscore

**Comment by zilzila** [comment]
Tables with leading underscore also work in my case:
```sql
create table _my_test_table (id int);

select * from _my_test_table d where d.
```

**Comment by anselal** [comment]
Your examples work for me too, but not for tables I already have.
Don't know what to do.
Will try to reinstall HeidiSQL and let you know if that fixed the issue

**Comment by anselal** [comment]
The issue still remains.
It works in previous versions though, like stable 10.3

**Comment by zilzila** [comment]
Thank you!
The problem is that I cannot reproduce this issue...

**Comment by anselal** [comment]
How does autocomplete work under the hood? Does it lookup `information_schema`? I would think that my `information_schema` got corrupted but as I mentioned it works under previous versions of HeidiSQL.

**Comment by ansgarbecker** [comment]
The IS schema is queried when you click a database or a table. The auto-completion just uses the details from these queries.

With the new logic for #914, I can also reproduce non-working completion, e.g. with comments:

```sql
SELECT *
FROM mytable s -- comment
WHERE s.<< nothing pops up
```

**Comment by anselal** [comment]
I just noticed that autocomplete works without a table alias

```
SELECT *
FROM mytable
WHERE mytable. << it pops up
```

**Comment by yphastos** [comment]
I don't know if this may help...

I have the same problem with autocompleting column names for tables, and noticed that in the log, the information from `information_schema`, and a few other queries like `show function status...` `show procedure status...` etc is being attempted to obtain with the schema name (database name) in lowercase, instead of the correct case, eg.

My DB name is `DEVEFERH` but is being searched as `deveferh` in lowercase, generating a 1049 error:

![image](https://user-images.githubusercontent.com/13773388/75565734-1c5e2880-5a14-11ea-99c1-30b167e43d4f.png)

I also noticed that autocompleting the TABLE names (instead of columns) does work, because in such case, the DB name seemingly is queried correctly, with the correct letter case

I hope that can help to solve the issue ASAP

**Comment by zilzila** [comment]
Yes, at the moment there is an incorrect handling of case-sensitive identifiers of MySql (with lower_case_table_names = 0). Now I'm working on this issue.

**Comment by zilzila** [comment]
I made fixes and pull request #920. Please report the test results.

**Comment by ansgarbecker** [comment]
I just merged your pull request, next nightly build will be there in a few minutes.

**Comment by anselal** [comment]
The issue still persist in 10.3.0.5903

**Comment by zilzila** [comment]
In order to find out where the error is and try to reproduce it, it would be nice to have data about the structure.
I found a way to solve the problem with private data! :) Replace all identifiers with aaa, bbb, ccc, etc., do obfuscation:
MyTableName.Column_1 -> AaAaaaAaaa.Bbbbbb_1 and so on.
Substitutions for identifiers must be the same, so that in the query and in the `show create table ...` substitutions correspond to the same column.

Please send the following:
1. An exact query with a problem table, without removing spaces, underscores, other characters other than a-z, A-Z, comments, without deleting/making new line breaks, etc. Only obfuscation of identifiers.
2. Result of `show create table <table_name>;` (columns `Table` and `Create Table`, obfuscated).
3. The result of `show create database <db_name>;` (columns `Database` and `Create Database`, obfuscated).

**Comment by ansgarbecker** [comment]
Until we have an exact reproduction recipe, you could fix the other one from [my comment](https://github.com/HeidiSQL/HeidiSQL/issues/917#issuecomment-592456631), using a comment after the table alias.

**Comment by zilzila** [comment]
Yes, I’m doing it right now.
Unfortunally, TRegexpr of this version does not support non-capturing groups `(?: )`, This is very lacking and complicates the code.
I looked at the SynEdit repository, it also has an old version of TRegExpr, although on the TRegExpr developer site you can download a version where `(?: )` is already supported.
Is the SynEdit project already abandoned and not developing? The last commits were a few months ago, there are several pull requests and I did not notice any activity from the SynEdit developers ...
Look-ahead and look-behind modifiers `(?= )` and `(?! )` are also sorely lacking, but TRegExpr developer explicitly states that there is no support and is not expected.
Maybe it makes sense to look for another class/package/etc to handle regular expressions, with normal support for modern modifiers and functionality?

**Comment by ansgarbecker** [comment]
@zilzila I have that TRegExpr update on the roadmap (#819). But although I planned to do that update for the next release, I decided to postpone that to a later one, as the next one is very soon (just a few days left). So I'd like to roll back the relevant commits now, and redo it for the next release, so you have more time to test and to fix remaining issues. Ok?

**Comment by zilzila** [comment]
Ok!

Perhaps it is better to make a branch for autocompletion? Because, as it seems to me, there are quite large changes coming that can affect many program modules.
I am talking about supporting spaces and other "forbidden" characters in quoted database names, `Conn.Database`. I have not watched it yet, but I believe that it will affect almost every piece of the program :) Here it will be necessary to test everything very carefully.
Also in my task list there is a task to deal with single quotes so that they do not double in quoted identifiers.
And I also want to parse the subqueries and then remove them from the FROM clause and add both alias and subquery columns to the proposal list.

A separate branch will allow committing and testing without affecting the master branch. And as soon as it is ready, it will be possible to merge into the master branch.
I do not know which option is better, you decide.

P.S. Updating to the latest version of TRegExpr is great!

**Comment by ansgarbecker** [comment]
I'd like to get just the problem from #895 fixed. This should be fixable without such large changes.

**Comment by zilzila** [comment]
> With the new logic for #914, I can also reproduce non-working completion, e.g. with comments:
> 
> ```sql
> SELECT *
> FROM mytable s -- comment
> WHERE s.<< nothing pops up
> ```

I made a fix for comments in the FROM table section, [committed](https://github.com/zilzila/HeidiSQL/commit/d1e7dd78c66cec74fd7978ad6624a6160b9c6511) it to my fork. 
What should I do now, create a pull request, just wait, or something else?

**Comment by anselal** [comment]
@ansgarbecker thnx, seems it has been fixed now !!!

**Comment by ansgarbecker** [comment]
@zilzila I'd like to keep it more simple, as written above. The issue with case sensitivity should be fixed already.

**Comment by zilzila** [comment]
Sorry for not responding so long, there was a lot of work.

I would also like to have as few changes as possible, but starting to make changes, I saw that one clings to the other and for the full implementation of autocompletion it will still be necessary (IMO) to do the handling of quoted identifiers and make other changes, as I described above.
So I'm afraid that for the full implementation of autocompletion there will be a lot to change. So I decided to take up this and implement a new logic. But if you prefer to leave it as it is, then no problem.

I’ll write my additions/tests of the current implementation of autocompletion in #895
